lesson5-LINQ/Task1/LinqTask.cs

```diff
@@ -8,85 +10,114 @@ public static class LinqTask
-            IEnumerable<Customer> customers
-        )
+        public static IEnumerable<(Customer customer, DateTime dateOfEntry)> Linq5(IEnumerable<Customer> customers)
         {
-            throw new NotImplementedException();
+            var result = customers
+                .Where(c => c.Orders.Length > 0)
+                .Select(c => (c, c.Orders.Min(o => o.OrderDate)))
+                .OrderBy(cd => cd.Item2.Year)
+                .ThenBy(cd => cd.Item2.Month)
+                .ThenByDescending(cd => cd.c.Orders.Sum(o => o.Total))
+                .ThenBy(cd => cd.c.CompanyName);
+
+            return result;
         }
 
         public static IEnumerable<Customer> Linq6(IEnumerable<Customer> customers)
         {
-            throw new NotImplementedException();
+            var result = customers.Where(c =>
+                Regex.IsMatch(c.PostalCode, @"[A-Za-z]+") 
+                || !Regex.IsMatch(c.Phone, @"^\(.*?\)") 
+                || c.Region == string.Empty || c.Region == null
+                );
+
+            return result;
         }
 
         public static IEnumerable<Linq7CategoryGroup> Linq7(IEnumerable<Product> products)
         {
-            /* example of Linq7result
-
-             category - Beverages
-	            UnitsInStock - 39
-		            price - 18.0000
-		            price - 19.0000
-	            UnitsInStock - 17
-		            price - 18.0000
-		            price - 19.0000
-             */
-
-            throw new NotImplementedException();
+            var result = products.GroupBy(p => p.Category).Select(g => new Linq7CategoryGroup
+            {
+                Category = g.Key,
+                UnitsInStockGroup = g.GroupBy(c => c.UnitsInStock).Select(p => new Linq7UnitsInStockGroup
+                {
+                    Prices = g.Select(product => product.UnitPrice)
+                }).OrderBy(price => price)
+            });
+
+            return result;
         }
 
-        public static IEnumerable<(decimal category, IEnumerable<Product> products)> Linq8(
-            IEnumerable<Product> products,
-            decimal cheap,
-            decimal middle,
-            decimal expensive
-        )
+        public static IEnumerable<(decimal category, IEnumerable<Product> products)> Linq8(IEnumerable<Product> products, decimal cheap, decimal middle, decimal expensive)
         {
-            throw new NotImplementedException();
+            var result = products.GroupBy(p => new
+            {
+                isCheap = p.UnitPrice <= cheap,
+                isMiddle = p.UnitPrice <= middle & p.UnitPrice > cheap,
+                isExpensive = p.UnitPrice <= expensive & p.UnitPrice > middle
+            }).Select(g =>
+            {
+                if (g.Key.isExpensive)
+                    return (expensive, g.AsEnumerable());
+                return g.Key.isMiddle ? (middle, g.AsEnumerable()) : (cheap, g.AsEnumerable());
```

Comment:
you can use switch expression here

Reply:
maybe the ternary operator is better?

Reply:
you have repetetive g.AsEnumerable(), also ternary operator should be in one line so if you have multiple ternaries it's difficult to read such code